### PR TITLE
Automatize and modernize deployments

### PR DESCRIPTION
**Pull request by 4383**
These changes aim to avoid unnecessary workload by automatizing major parts of the release process.
    
With these changes, deployments would be automatic.
Pypi will accept new releases coming from eventlet/eventlet by adding a trusted publisher based on OIDC:
 - https://docs.pypi.org/trusted-publishers/
 - https://en.wikipedia.org/wiki/OpenID#OpenID_Connect_(OIDC)
    
These changes propose:
- Manage version number by using git tags rather than manual versioning (in python code)
- Modernize packaging by using pyproject.toml rather than setup.py
- Remove old artifacts related to publishing (`bin/release`)
- Adding a github workflow dedicated to pypi publish
- Added OIDC trusted publisher management https://pypi.org/manage/project/eventlet/settings/publishing/

These changes updated the list of authors to reflect current active core maintainers and allow community members    to contact the team.

    
Related to recent discussions:
- https://github.com/eventlet/eventlet/issues/843
- https://github.com/eventlet/eventlet/issues/842

### CODE REVIEW

**Comment by damani42** [comment]
Looks good to merge.

**Comment by 4383** [comment]
Please, let's wait a bit before merging this patch.

I'd appreciate to get @itamarst's feedback too, as we discussed this topic together last Friday.

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+dependencies = []
+
+[project.urls]
+Homepage = "https://github.com/eventlet/eventlet"
+History = "https://github.com/eventlet/aiohub/blob/main/NEWS"
+Tracker = "https://github.com/eventlet/eventlet/issues"
+Source = "https://github.com/eventlet/aiohub"
+
+[project.optional-dependencies]
+dev = ["black", "isort", "pip-tools", "build", "twine", "pre-commit", "commitizen"]
```
Comment (4383) [inline pyproject.toml:44]: This is optional development dependencies.

If you are ok we that, once major issues will be solved, we could start using `pre-commit` and migrate to `black` formatting. Using `black` would avoid to loose time with formatting issues and pep8...

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+requires = [
+  "hatch-vcs>=0.3",
+  "hatchling>=1.12.2",
+]
+build-backend = "hatchling.build"
+
+[project]
+name = "eventlet"
+authors = [
+    {name = "Liden Lab", email = "eventletdev@lists.secondlife.com"},
```
Comment (4383) [inline pyproject.toml:11]: Wonder if we want to update the author, I'm not sure this email address still work and I don't know if someone among us still own access to this inbox, any idea?

Maybe we could create a mailing list to allow user to contact us and then use the address of this mailing list there.
Also I'd suggest replacing the author name by the eventlet orga.

I don't want to hide the previous author, so, we could keep this line in the authors, and, then, just add another line before, to reflect the current maintainers information.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        python-version: '3.x'
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install build
+    - name: Build package
+      run: python -m build
+    - name: Publish package
+      # deploy only when a new tag is pushed to github
+      if: github.event_name == 'push' && startsWith(github.ref, 'refs/tags')
```
Comment (4383) [inline .github/workflows/publish.yml:37]: publishing on pypi would be allowed only when we push a new tag.

**Comment by 4383** [comment]
the pypy jobs seems a bit flaky... the error doesn't seems legit, but, I'm not able to trigger an other run.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    'greenlet >= 1.0',
+    'monotonic >= 1.4;python_version<"3.5"',
+    'six >= 1.10.0',
+]
+
+[project.urls]
+Homepage = "https://github.com/eventlet/eventlet"
+History = "https://github.com/eventlet/aiohub/blob/main/NEWS"
+Tracker = "https://github.com/eventlet/eventlet/issues"
+Source = "https://github.com/eventlet/aiohub"
```
Comment (itamarst) [inline pyproject.toml:46]: Wrong URL

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+dependencies = [
+    'dnspython >= 1.15.0',
+    'greenlet >= 1.0',
+    'monotonic >= 1.4;python_version<"3.5"',
+    'six >= 1.10.0',
+]
+
+[project.urls]
+Homepage = "https://github.com/eventlet/eventlet"
+History = "https://github.com/eventlet/aiohub/blob/main/NEWS"
```
Comment (itamarst) [inline pyproject.toml:44]: Wrong URL

**Comment by itamarst** [comment]
This all seems fine, other than:

* A few minor comments, see above.
* Deciding #846 maybe.

**Comment by itamarst** [comment]
I guess merging this doesn't actually mean doing a release, and #836 is a blocker for a release, not release infrastructure. So please address and merge, thank you!

**Review by itamarst — APPROVED** [review]

**Comment by 4383** [comment]
done

**Comment by 4383** [comment]
> I guess merging this doesn't actually mean doing a release, and #836 is a blocker for a release, not release infrastructure. So please address and merge, thank you!

These changes won't release a new version, they are just the machinery.
We can merge this patch, manage #836, and then release something.

**Comment by 4383** [comment]
We are now ready to bring back Python 3.7 at life

File: eventlet/__init__.py [hunk 6]
```diff
@@ -8,71 +8,65 @@
         DeprecationWarning,
     )
 
-version_info = (0, 33, 3)
-__version__ = '.'.join(map(str, version_info))
```
[the remark this answers is not in the document]
Comment (4383) [reply eventlet/__init__.py:12]: Yes I removed it intentionally.
Now versions are managed by using code version control (git tags).

Version can be retrieved by using pip.

Someone else reported a similar issue yesterday evening, see the discussion there: https://github.com/eventlet/eventlet/issues/842#issuecomment-1864896273

I proposed to reintroduced something like:

```python
from importlib.metadata import version as getversion

__version__ = getversion("eventlet")
```

However, I'd love to understand why gunicorn is checking versions like that and not by using requirements file or dependencies mechanisms. Any idea?

[the remark this answers is not in the document]
Comment (jstasiak) [reply eventlet/__init__.py:12]: I'm not gunicorn but speaking from my experience:

Sometimes you need to support multiple versions and modify your behavior based on the version you deal with.

Sometimes you want to import a package from a development repository tarball you just downloaded (where there's no metadata or Git tags present).

[the remark this answers is not in the document]
Comment (4383) [reply eventlet/__init__.py:12]: Fix proposed https://github.com/eventlet/eventlet/pull/859

[the remark this answers is not in the document]
Comment (4383) [reply eventlet/__init__.py:12]: > Sometimes you want to import a package from a development repository tarball you just downloaded (where there's no metadata or Git tags present).

That confirms my choice to not use `setuptools_scm `or `importlib.metadata.version`

**Pull request merged**